pulsar-common/src/main/java/org/apache/pulsar/client/api/url/DataURLStreamHandler.java

```diff
@@ -57,20 +60,23 @@ public void connect() throws IOException {
             if (this.uri == null) {
                 throw new IOException();
             }
-            Pattern pattern = Pattern.compile(
-                    "(?<mimeType>.+?)(;(?<charset>charset=.+?))?(;(?<base64>base64?))?,(?<data>.+)", Pattern.DOTALL);
+
             Matcher matcher = pattern.matcher(this.uri.getSchemeSpecificPart());
             if (matcher.matches()) {
                 this.contentType = matcher.group("mimeType");
-                String charset = matcher.group("charset");
-                if (charset == null) {
-                    charset = "US-ASCII";
+                if (contentType == null) {
+                    this.contentType = "application/data";
                 }
+
+                for (int i =0; i < matcher.groupCount(); i++) {
+                    System.out.println("Group: " + matcher.group(i));
+                }
+
                 if (matcher.group("base64") == null) {
                     // Support Urlencode but not decode here because already decoded by URI class.
-                    this.data = new String(matcher.group("data").getBytes(), charset);
+                    this.data = matcher.group("data").getBytes();
```

Comment:
Well, this is probably equivalent to the original code, so this change is fine.

However, I realized that the both code have the same issue. Because we drop charset information here, users cannot read the data as a string with the specified charset. I think we should stop parsing parameters in mimetype and return the whole mimetype part as content-type.

Reply:
The problem with original code is that it was getting a list of bytes and converting them into a `String` (which in case of keys is not possible) and then exposing back to a `ByteArrayInputStream`. I just cut through the String conversion

Reply:
Yeah, I can see that, and we shoud fix that part. So this would be out of this PR’s scope but we need to return charset as a part of contentType so that URL class users can convert the data into String after reading the raw data from InputStream. I think we can do this by removing charset part from the regex.